### ISSUE TITLE
Feature request: option to unsoundly ignore more errors issued because of #979

### ISSUE THREAD

**Issue by kelloggm** [typetools/checker-framework#3001]
False positives for checkers that only issue a few errors (like the [Object Construction Checker](https://github.com/kelloggm/object-construction-checker)) are dominated by errors caused by #979. When running these checkers, suppressing these warnings is time-consuming and inconvenient for users, because they are unrelated to the property or properties actually being checked.

Some errors related to #979 are already [unsoundly suppressed](https://github.com/typetools/checker-framework/issues/979#issuecomment-279578221) by default. This feature request is to suggest that we do something similar for another class of errors. Whether or not the default is the conservative behavior isn't important to me.

The errors arise from programs like the following:

```java
import java.lang.Class;
import com.sun.tools.javac.code.Symbol;
import com.sun.tools.javac.code.Symbol.ClassSymbol;
import javax.lang.model.element.ElementKind;
import com.sun.tools.javac.code.Scope;
import com.sun.tools.javac.code.Type;
import com.sun.tools.javac.code.Types;
import com.sun.tools.javac.util.Name;

class Issue979RemainingError {
    private final Types types = null;

    private <T extends Symbol> T getMember(
      Class<T> type, ElementKind kind, Symbol classSymbol, Name name) {
	if (classSymbol.type == null) {
	    return null;
	}
	for (Type t : types.closure(classSymbol.type)) {
	    Scope scope = t.tsym.members();
	    for (Symbol sym : scope.getSymbolsByName(name)) {
		if (sym.getKind().equals(kind)) {
		    return type.cast(sym);
		}
	    }
	}
	if (classSymbol.hasOuterInstance()) {
	    T sym = getMember(type, kind, classSymbol.type.getEnclosingType().asElement(), name);
	    if (sym != null) {
		return sym;
	    }
	}
	if (classSymbol.owner != null
	    && classSymbol != classSymbol.owner
	    && classSymbol.owner instanceof Symbol.ClassSymbol) {
	    T sym = getMember(type, kind, classSymbol.owner, name);
	    if (sym != null && sym.isStatic()) {
		return sym;
	    }
	}
	return null;
    }
}
```

Typechecking this class with e.g. the Value Checker, which should issue no errors, does issue two errors.

The command I ran:
```
javac -processor org.checkerframework.common.value.ValueChecker Issue979RemainingErrors.java 
```

The output:
```
Issue979RemainingErrors.java:29: error: [return.type.incompatible] incompatible types in return.
		return sym;
		       ^
  type of expression: T[ extends @UnknownVal Symbol super @UnknownVal Void]
  method return type: T[ extends @UnknownVal Symbol super @BottomVal Void]
Issue979RemainingErrors.java:37: error: [return.type.incompatible] incompatible types in return.
		return sym;
		       ^
  type of expression: T[ extends @UnknownVal Symbol super @UnknownVal Void]
  method return type: T[ extends @UnknownVal Symbol super @BottomVal Void]
2 errors
```

Note that in each case, the error has some similar properties:
1. the type of the expression is `T extends <Top> super <Top>`
2. the required type is `T extends <Top> super <Bottom>`

Running this example through another type system that should issue no errors, such as the Tainting Checker, shows similar behavior:

```
Issue979RemainingErrors.java:29: error: [return.type.incompatible] incompatible types in return.
		return sym;
		       ^
  type of expression: T[ extends @Tainted Symbol super @Tainted Void]
  method return type: T[ extends @Tainted Symbol super @Untainted Void]
Issue979RemainingErrors.java:37: error: [return.type.incompatible] incompatible types in return.
		return sym;
		       ^
  type of expression: T[ extends @Tainted Symbol super @Tainted Void]
  method return type: T[ extends @Tainted Symbol super @Untainted Void]
2 errors
```

I propose that we implement a check to prevent such errors from being issued, based on the pattern of the required type being `<GenericType> extends <Top> super <Bottom>` when the actual type is `<GenericType> extends <Top> super <Top>`.

I'm happy to do the implementation work, but I'm not exactly sure where this information is available that I'd be able to prevent the error from being issued. I've considered two options:
* in `SourceChecker#report`, but the top and bottom types in the qualifier hierarchy aren't available there (or at least, I don't know how to get to them)
* in the common assignment check, but not all errors go through the CAC. That still might be the best way to do it

@smillst, do you have any other ideas on how I should do this, or on which approach you would take?

@msridhar suggested that I do this in public so he could see our discussion rather than just sending an email, so I'm pinging him

**Comment by smillst** [comment]
This example isn't Java 8 type inference, so the existing algorithm should handle it.  I was able to find and fix the bug (#3003).  Do you have another example that doesn't pass with #3003?

**Comment by msridhar** [comment]
I can try to test with one of our larger Object Construction Checker examples.  Let me see what I find.

**Comment by msridhar** [comment]
This is taking longer for me to test than I anticipated.  @kelloggm if you have some easy test analysis / benchmark to see if this problem has gone away please go ahead.

**Comment by msridhar** [comment]
Ok, I retried on a large object construction benchmark and I'm still seeing this issue.  Here is one message that I think matches the `super` pattern @kelloggm mentioned:

```
/Users/msridhar/git-repos/dljc-experiment/outdir/nomulus/core/src/main/java/google/registry/beam/invoicing/InvoicingPipeline.java:154: error: [argument.type.incompatible] incompatible types in argument.
                  .via(BillingEvent::getInvoiceGroupingKey))
                      ^
  found   : @CalledMethodsTop MapElements</*INFERENCE FAILED for:*/ ? extends @CalledMethodsTop Object, /*INFERENCE FAILED for:*/ ? extends @CalledMethodsTop Object>
  required: @CalledMethodsTop PTransform<? extends @CalledMethodsTop PInput super @CalledMethodsTop PCollection<@CalledMethodsTop BillingEvent>, /*INFERENCE FAILED for:*/ @CalledMethodsTop ? extends @CalledMethodsTop POutput>
```

[Here](https://github.com/msridhar/nomulus/blob/base/core/src/main/java/google/registry/beam/invoicing/InvoicingPipeline.java#L154) is the relevant line (I haven't fully understood what is going on).

Here's another one:

```
/Users/msridhar/git-repos/dljc-experiment/outdir/nomulus/core/src/main/java/google/registry/tools/GenerateDnsReportCommand.java:105: error: [return.type.incompatible] incompatible types in return.
                      ImmutableMap.of(
                                     ^
  type of expression: @CalledMethodsTop ImmutableMap<@CalledMethodsTop String, @CalledMethodsTop Serializable & @CalledMethodsTop Comparable<?[ extends @CalledMethodsTop Serializable & @CalledMethodsTop Comparable<?[ extends @CalledMethodsTop Object super @CalledMethodsTop Void]> super @CalledMethodsTop Void]>>
  method return type: @CalledMethodsTop ImmutableMap<@CalledMethodsTop String, @CalledMethodsTop Serializable & @CalledMethodsTop Comparable<?[ extends @CalledMethodsTop Serializable & @CalledMethodsTop Comparable<?[ extends @CalledMethodsTop Object super @CalledMethodsBottom Void]> super @CalledMethodsBottom Void]>>
```

[Relevant line](https://github.com/msridhar/nomulus/blob/base/core/src/main/java/google/registry/tools/GenerateDnsReportCommand.java#L105)

If you need a repro test case let me know, but that will take some time to make.

**Comment by kelloggm** [comment]
Re-opening, because we believe this issue is broader than just what #3003 addressed.

**Comment by smillst** [comment]
> I propose that we implement a check to prevent such errors from being issued, based on the pattern of the required type being `<GenericType> extends <Top> super <Bottom>` when the actual type is `<GenericType> extends <Top> super <Top>`.

This would suppress many true warnings unrelated to type argument inference and, given the errors @msridhar reported, it wouldn't suppress all the warnings you want it to suppress.

I would suggest adding an option that would cause the BaseTypeVisitor to not check any expression requiring Java 8 type argument inference.

**Comment by kelloggm** [comment]
I tried to follow @smillst's suggestion above to add an option that would cause BaseTypeVisitor to not typecheck any expression requiring Java 8 type inference. I'm writing this comment to document what problems I ran into and what I learned about what might be causing these false positives. It's also possible I misunderstood @smillst's suggestion.

`DefaultTypeHierarchy` already by default ignores "raw" types---those without explicit generic arguments, which therefore require inference---when computing subtyping relationships. That suggests that adding another flag to BaseTypeVisitor to do the same thing isn't the right choice, because it ought to be handled by the existing machinery. (Or was @smillst referring to something other than raw types with her suggestion above?)

I suspect that the first error @msridhar reported above is caused by a bug in the implementation of the work-around for #802. Without a reproducible test case, however, it's impossible to say for sure.

The second error looks a lot like #1032, and quite similar to the original bug I reported here. Again, with a reproducible test case I'm sure we could debug it and fix whatever is wrong in the inference code that's causing it.

My conclusion from looking through all this code is that what we want is already implemented, for the most part, but that there are bugs and/or un-implemented features. If we want to fix them, we need to produce reproducible test cases for them individually and then go from there; there's not going to be a silver bullet we can write that will filter them all out.

**Comment by msridhar** [comment]
A reduced test case for the first error I reported is below (probably can be further reduced).  Hopefully this will help track things down.  Let me know if I should open a separate issue.

```java
import java.io.Serializable;

public class InvoicingPipeline {

  public static class PCollection<T> implements PValue {

    public <OutputT extends POutput> OutputT apply(
        String name, PTransform<? super PCollection<T>, OutputT> t) {
      return null;
    }
  }

  public abstract static class PTransform<InputT extends PInput, OutputT extends POutput> {
    public abstract void expand(InputT input);
  }

  static interface PInput {}

  static interface POutput {}

  interface PValue extends PInput, POutput {}

  static class BillingEvent {

    InvoiceGroupingKey getInvoiceGroupingKey() {
      return null;
    }
  }

  public static class MapElements<InputT, OutputT>
      extends PTransform<PCollection<? extends InputT>, PCollection<OutputT>> {
    public static <OutputT> MapElements<?, OutputT> into(final TypeDescriptor<OutputT> outputType) {
      return null;
    }

    public static <InputT, OutputT> MapElements<InputT, OutputT> via(
        final ProcessFunction<InputT, OutputT> fn) {
      return null;
    }

    public void expand(PCollection<? extends InputT> input) {}
  }

  static class TypeDescriptor<T> {
    public static <T> TypeDescriptor<T> of(Class<T> type) {
      return null;
    }
  }

  static class InvoiceGroupingKey {}

  @FunctionalInterface
  public interface ProcessFunction<InputT, OutputT> extends Serializable {
    OutputT apply(InputT input) throws Exception;
  }

  private static class GenerateInvoiceRows
      extends PTransform<PCollection<BillingEvent>, PCollection<String>> {
    @Override
    public void expand(PCollection<BillingEvent> input) {
      input.apply(
          "Map to invoicing key",
          MapElements.into(TypeDescriptor.of(InvoiceGroupingKey.class))
              .via(BillingEvent::getInvoiceGroupingKey));
    }
  }
}

```

**Comment by msridhar** [comment]
I confirmed that when I copy the above input into `framework/tests/all-systems` I get test failures, e.g., in `UnitsTest`.

**Comment by smillst** [comment]
>DefaultTypeHierarchy already by default ignores "raw" types---those without explicit generic arguments, which therefore require inference---when computing subtyping relationships. 

 A raw type is a type in Java source code that doesn't have type arguments.  Although their type arguments may be inferred by the Checker Framework, this isn't the sort of inference that #979 is referring.  It's referring to implementing inference for method type arguments.  In the case of that method type argument inference fails, the type arguments are initialized to wildcards with the uninferedTypeArgument field set to true.  By default, `isSubtype` returns true when it finds an an uninferred type argument. 

 But as you have seen, just return true in `isSubtype` is not enough.   I was suggesting that the BaseTypeVisitor should not visit expressions that require type Java 8 type argument inference.  I'm looking into how to determine which expressions these are.

**Comment by kelloggm** [comment]
I reduced the other failure @msridhar reported, which I'm copying below for completeness. I also opened a new issue.

```
import java.util.List;
import java.util.Set;
import java.util.Map;
import java.util.HashMap;
import java.util.stream.Collectors;

public class LambdaInStream {

    public Set<MyInnerClass> getDsData() {
        return null;
    }

    public static class MyInnerClass {
        public int getKeyTag() {
            return 5;
        }

        public String getDigest() {
            return null;
        }
    }

    private void write() {
        List<Map<String, ?>> dsData =
                        getDsData()
                        .stream()
                        .map(dsData1 ->
                                        ImmutableMap.of(
                                                "keyTag", dsData1.getKeyTag(),
                                                "digest", dsData1.getDigest()))
                        .collect(Collectors.toList());
    }

    public static class ImmutableMap<K, V> extends HashMap<K, V> {
        public static <K, V> ImmutableMap<K, V> of(K k1, V v1, K k2, V v2) {
            return null;
        }
    }
}
```

**Comment by msridhar** [comment]
Small update here: on a local build with #3035 and #3042 applied, I do not see any more generics-inference-related errors when running the object construction checker on nomulus.
